x/wasm/keeper/handler_plugin.go

```diff
@@ -161,54 +163,115 @@ func (m MessageHandlerChain) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAd
-	if msg.IBC == nil || msg.IBC.SendPacket == nil {
+	if msg.IBC == nil {
 		return nil, nil, nil, types.ErrUnknownMsg
 	}
-	if contractIBCPortID == "" {
-		return nil, nil, nil, errorsmod.Wrapf(types.ErrUnsupportedForContract, "ibc not supported")
-	}
-	contractIBCChannelID := msg.IBC.SendPacket.ChannelID
-	if contractIBCChannelID == "" {
-		return nil, nil, nil, errorsmod.Wrapf(types.ErrEmpty, "ibc channel")
-	}
+	switch {
+	case msg.IBC.SendPacket != nil:
+		if contractIBCPortID == "" {
+			return nil, nil, nil, errorsmod.Wrapf(types.ErrUnsupportedForContract, "ibc not supported")
+		}
+		contractIBCChannelID := msg.IBC.SendPacket.ChannelID
+		if contractIBCChannelID == "" {
+			return nil, nil, nil, errorsmod.Wrapf(types.ErrEmpty, "ibc channel")
+		}
 
-	channelCap, ok := h.capabilityKeeper.GetCapability(ctx, host.ChannelCapabilityPath(contractIBCPortID, contractIBCChannelID))
-	if !ok {
-		return nil, nil, nil, errorsmod.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
-	}
-	seq, err := h.ics4Wrapper.SendPacket(ctx, channelCap, contractIBCPortID, contractIBCChannelID, ConvertWasmIBCTimeoutHeightToCosmosHeight(msg.IBC.SendPacket.Timeout.Block), msg.IBC.SendPacket.Timeout.Timestamp, msg.IBC.SendPacket.Data)
-	if err != nil {
-		return nil, nil, nil, errorsmod.Wrap(err, "channel")
-	}
-	moduleLogger(ctx).Debug("ibc packet set", "seq", seq)
+		channelCap, ok := h.capabilityKeeper.GetCapability(ctx, host.ChannelCapabilityPath(contractIBCPortID, contractIBCChannelID))
+		if !ok {
+			return nil, nil, nil, errorsmod.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
+		}
+		seq, err := h.ics4Wrapper.SendPacket(ctx, channelCap, contractIBCPortID, contractIBCChannelID, ConvertWasmIBCTimeoutHeightToCosmosHeight(msg.IBC.SendPacket.Timeout.Block), msg.IBC.SendPacket.Timeout.Timestamp, msg.IBC.SendPacket.Data)
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "channel")
+		}
+		moduleLogger(ctx).Debug("ibc packet set", "seq", seq)
 
-	resp := &types.MsgIBCSendResponse{Sequence: seq}
-	val, err := resp.Marshal()
-	if err != nil {
-		return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
-	}
-	any, err := codectypes.NewAnyWithValue(resp)
-	if err != nil {
-		return nil, nil, nil, errorsmod.Wrap(err, "failed to convert IBC send response to Any")
+		resp := &types.MsgIBCSendResponse{Sequence: seq}
+		val, err := resp.Marshal()
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
+		}
+		any, err := codectypes.NewAnyWithValue(resp)
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to convert IBC send response to Any")
+		}
+		msgResponses := [][]*codectypes.Any{{any}}
+
+		return nil, [][]byte{val}, msgResponses, nil
+	case msg.IBC.WriteAcknowledgement != nil:
```

Comment:
I'm assuming only the contract itself can send this message

Reply:
Yes, this is part of the MessageDispatcher chain that is used to execute messages returned by a contract entrypoint. It's not a normal sdk message and the ibc port id that is used here is the one of the contract who sent this.